File: saga/bench/bench_tokenizers.ml
(* Benchmark suite for Saga tokenizers using realistic fixtures. *)

open Saga_tokenizers

module Fixtures = struct
  let data_dir = Filename.concat (Sys.getcwd ()) "saga/bench/data"

  let read_file name =
    let path = Filename.concat data_dir name in
    let ic = open_in_bin path in
    let len = in_channel_length ic in
    let content = really_input_string ic len in
    close_in ic;
    content

  let load_tokenizer name =
    let path = Filename.concat data_dir name in
    match Tokenizer.from_file path with
    | Ok tok -> tok
    | Error err ->
        failwith
          (Printf.sprintf "Failed to load tokenizer %s: %s" path
             (Printexc.to_string err))

  let short_text = read_file "news_1k.txt"
  let long_text = read_file "wiki_64k.txt"

  let batch_32 =
    let rec loop acc remaining =
      if remaining = 0 then List.rev acc
      else loop (short_text :: acc) (remaining - 1)
    in
    loop [] 32
end

let encode_single tok text = Tokenizer.encode tok text |> ignore
let encode_batch tok texts = Tokenizer.encode_batch tok texts |> ignore
let decode_ids tok ids = Tokenizer.decode tok ids |> ignore

let make_suite ~label ~tokenizer =
  let open Fixtures in
  let decode_input =
    let encoding = Tokenizer.encode tokenizer long_text in
    Array.copy (Encoding.get_ids encoding)
  in
  let benches =
    [
      Ubench.bench "Encode/single_short" (fun () ->
          encode_single tokenizer short_text);
      Ubench.bench "Encode/single_long" (fun () ->
          encode_single tokenizer long_text);
      Ubench.bench "Encode/batch_32" (fun () -> encode_batch tokenizer batch_32);
      Ubench.bench "Decode/long" (fun () -> decode_ids tokenizer decode_input);
    ]
  in
  Ubench.group label benches

let all_benchmarks =
  let open Fixtures in
  let bpe =
    make_suite ~label:"BPE" ~tokenizer:(load_tokenizer "byte_bpe.json")
  in
  let wordpiece =
    make_suite ~label:"WordPiece" ~tokenizer:(load_tokenizer "wordpiece.json")
  in
  let wordlevel =
    make_suite ~label:"WordLevel" ~tokenizer:(load_tokenizer "wordlevel.json")
  in
  [ bpe; wordpiece; wordlevel ]

let () = Ubench.run_cli all_benchmarks

File: saga/bench/bench_tokenizers.py
from __future__ import annotations

from pathlib import Path
from typing import Any, Callable, List

from tokenizers import Tokenizer

_ROOT = Path(__file__).resolve().parent
_DATA_DIR = _ROOT / "data"

import sys

vendor_dir = _ROOT.parents[1] / "vendor" / "ubench"
if str(vendor_dir) not in sys.path:
    sys.path.insert(0, str(vendor_dir))

import ubench  # type: ignore


SHORT_TEXT = (_DATA_DIR / "news_1k.txt").read_text(encoding="utf-8")
LONG_TEXT = (_DATA_DIR / "wiki_64k.txt").read_text(encoding="utf-8")
BATCH_32 = [SHORT_TEXT] * 32


def load_tokenizer(filename: str) -> Tokenizer:
    path = _DATA_DIR / filename
    return Tokenizer.from_file(str(path))


def make_suite(label: str, tokenizer: Tokenizer) -> Any:
    decode_ids = tokenizer.encode(LONG_TEXT).ids

    benches: List[Any] = [
        ubench.bench("Encode/single_short", lambda: tokenizer.encode(SHORT_TEXT)),
        ubench.bench("Encode/single_long", lambda: tokenizer.encode(LONG_TEXT)),
        ubench.bench("Encode/batch_32", lambda: tokenizer.encode_batch(BATCH_32)),
        ubench.bench("Decode/long", lambda: tokenizer.decode(decode_ids)),
    ]

    return ubench.group(label, benches)


def build_benchmarks() -> List[Any]:
    return [
        make_suite("GPT-2", load_tokenizer("gpt2.json")),
        make_suite("BERT-base", load_tokenizer("bert_base.json")),
        make_suite("LLaMA", load_tokenizer("llama.json")),
    ]


def default_config() -> ubench.Config:
    return ubench.Config.default().build()


def main() -> None:
    benchmarks = build_benchmarks()
    config = default_config()
    ubench.run(benchmarks, config=config, output_format="pretty", verbose=False)


if __name__ == "__main__":
    main()
